core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -209,6 +214,27 @@ public UfsClient getRoot() {
     }
   }
 
+  @Override
+  public UfsClient getJournal(URI location) {
+    synchronized (this) {
+      if (mJournalUfsClientMap == null) {
+        mJournalUfsClientMap = new HashMap<>();
+      }
+      UfsClient ufsClient = mJournalUfsClientMap.get(location.getScheme());
```

Comment:
I am confused. Why could there be multiple journal locations?

Shouldn't this look similar to the root ufs client, which there is only 1 of?

Reply:
From the master.log, It seems that there are multiple journal for different service need format.

```
2020-12-16 10:48:04,744 INFO  cli.Format (Format.java:format(121)) - Formatting master journal: /Users/mbl/projects/github/alluxio-2/journal/
2020-12-16 10:48:04,825 INFO  extensions.ExtensionFactoryRegistry (ExtensionFactoryRegistry.java:scanLibs(169)) - Loading core jars from /Users/mbl/projects/github/alluxio-2/lib
2020-12-16 10:48:05,000 INFO  extensions.ExtensionFactoryRegistry (ExtensionFactoryRegistry.java:scanExtensions(159)) - Loading extension jars from /Users/mbl/projects/github/alluxio-2/extensions
2020-12-16 10:48:05,032 INFO  ufs.UfsJournal (UfsJournal.java:format(428)) - Formatting /Users/mbl/projects/github/alluxio-2/journal/BlockMaster/v1
2020-12-16 10:48:05,058 INFO  ufs.UfsJournal (UfsJournal.java:format(428)) - Formatting /Users/mbl/projects/github/alluxio-2/journal/TableMaster/v1
2020-12-16 10:48:05,060 INFO  ufs.UfsJournal (UfsJournal.java:format(428)) - Formatting /Users/mbl/projects/github/alluxio-2/journal/FileSystemMaster/v1
2020-12-16 10:48:05,060 INFO  ufs.UfsJournal (UfsJournal.java:format(428)) - Formatting /Users/mbl/projects/github/alluxio-2/journal/MetaMaster/v1
2020-12-16 10:48:05,061 INFO  ufs.UfsJournal (UfsJournal.java:format(428)) - Formatting /Users/mbl/projects/github/alluxio-2/journal/MetricsMaster/v1
2020-12-16 10:48:05,062 INFO  cli.Format (Format.java:main(106)) - Formatting complete
```

Also, from the following Format#format, there are lots of master services, each one need create journal.

```java

public static void format(Mode mode, AlluxioConfiguration alluxioConf) throws IOException {
    NoopUfsManager noopUfsManager = new NoopUfsManager();
    switch (mode) {
      case MASTER:
        URI journalLocation = JournalUtils.getJournalLocation();
        LOG.info("Formatting master journal: {}", journalLocation);
        JournalSystem journalSystem = new JournalSystem.Builder()
            .setLocation(journalLocation).build(CommonUtils.ProcessType.MASTER);
        for (String masterServiceName : ServiceUtils.getMasterServiceNames()) {
          journalSystem.createJournal(new NoopMaster(masterServiceName, noopUfsManager));
        }
        journalSystem.format();
        break;
...
```

Reply:
@maobaolong I see. Then, does this PR actually help with the original motivation of avoiding loading the UFS journal jars? It seems like this will create a new UFS client for each master service (block, fs, table, etc) to format each, and each one will have a different location, so it will load the jars again?

In this example, I feel like the ideal behavior should be to load the UFS jars once for the location `/Users/mbl/projects/github/alluxio-2/journal/`, and all the master services are going to be sub-directories of that single UFS journal location. Does this current PR already behave in this way?

Reply:
@gpang Thanks for your review again. I think it just load the jars once during the format, since `location.getScheme()`, I use the scheme of the location as the key mapping to the related Underfilesystem.

And I did a format test using this PR, the behavior running as expected.

Reply:
@maobaolong O, you are right. the scheme will always be the same. Then I am confused as to why we even need the map to begin with. The scheme for all the journals should always be the same, since they are always a sub-directory of the journal location right?

I feel like the map will always have only 1 entry, so we should just not use a map?

Reply:
@gpang OK, I remove the map now.